### ISSUE TITLE
CJS Require fails with Cannot find module

### ISSUE THREAD

**Issue by sjustintaylor** [panva/jose#105]
**Describe the bug**
CJS import fails with " Cannot find module 'jose/jwt/verify'" when trying to import the jwtVerify module in a jest test file. 
```
    Cannot find module 'jose/jwt/verify' from 'modules/tokens.test.js'

    > 1 | const { default: jwtVerify } = require("jose/jwt/verify");

      at Resolver.resolveModule (node_modules/jest-resolve/build/index.js:306:11)
      at Object.<anonymous> (modules/tokens.test.js:1:32)
```

**To Reproduce**
I've created a repo that exhibits the same behavior [here](https://github.com/sjustintaylor/jose-issue). It exhibits this behavior on both my desktop and my work laptop (both running ubuntu).
I'm running `yarn test` inside the server directory when the error occurs.

Code to reproduce the behavior:
Inside a issue.test.js file, in `server/modules`:
```js
const { default: jwtVerify } = require("jose/jwt/verify");
```

**Expected behaviour**
The module is imported and ready for use.

**Environment:**

- `jose` version: v3.0.2 (also tried 3.0.0, and 3.0.1)
- affected runtime is: Node.js 14.12.0, 14.15.0, 12.19.0, 13.14 (I tried all 4, same result)
- OS: Ubuntu 20.04
- nvm: 0.37.0
- yarn: 1.22.5

**Additional context**
I can `require` other modules (I tried express, and dotenv) into the test file without a problem. Direct imports (`require('../node_modules/jose/dist/node/cjs/jwt/verify')`) also fail, with ` ReferenceError: TextEncoder is not defined`.
It's being used in an express server, with both the client and the server in the same repo (`./server` and `./client` in the same directory). The commands for the server are always run inside the server directory. I've installed @types/node, and @types/jest, to no effect. I've also tried deleting and recreating the `yarn.lock` file, reinstalling dependencies, and cleaning the yarn cache before reinstalling.

- [ X] the bug is happening on latest `jose` too.
- [ X] i have searched the issues tracker on github for similar issues and couldn't find anything related.

**Comment by panva** [comment]
When I simply run `node modules/tokens.test.js` (after having commented out the describe block) the require works just fine. That, plus the actual error stack indicates this is a problem with jest's resolve module. It doesn't actually support the package "exports" field.

https://github.com/facebook/jest/issues/9430
https://github.com/facebook/jest/issues/9771

I think you may be able to use jest's [custom resolver feature](https://jestjs.io/docs/en/configuration#resolver-string) to work around this deficiency in jest.

**Comment by panva** [comment]
Now as far as `ReferenceError: TextEncoder is not defined` goes, that's a jest problem too - https://github.com/facebook/jest/issues/2549


The TextEncoder and TextDecoder classes are available on the global object since node v11.0.0 and jest strips all of that to inject an isolated context.

**Comment by sjustintaylor** [comment]
Thank you for your help, and the quick response. I'll take a look at the custom resolver (or migrate to mocha).

**Comment by panva** [comment]
> or migrate to mocha

🙏 : https://github.com/avajs/ava

**Comment by jschmold** [comment]
#110 
Referencing a solution for @sjustintaylor that I found

**Comment by matthieubosquet** [comment]
> The TextEncoder and TextDecoder classes are available on the global object since node v11.0.0 and jest strips all of that to inject an isolated context. Using https://www.npmjs.com/package/jest-environment-node-single-context helps there.

@sjustintaylor adding `testEnvironment: "node"` to my `jest.config.js` fixed the TextEncoder error. I found mention of this parameter in the [Jest 26 release notes](https://jestjs.io/blog/2020/05/05/jest-26).

Following @jschmold 's [suggestion](https://github.com/panva/jose/issues/110#issuecomment-731331109) and using `moduleNameMapper` to fix the import error worked.

Here is my sample fully working jest config:

```javascript
module.exports = {
  preset: "ts-jest",
  testEnvironment: "node",
  moduleNameMapper: {
    "^jose/(.*)$": "<rootDir>/node_modules/jose/dist/node/cjs/$1"
  },
};
```

[Module name mapper configuration in Jest's documentation](https://jestjs.io/docs/en/configuration#modulenamemapper-objectstring-string--arraystring).
